README.md
# Stripe Python bindings

## Installation

You don't need this source code unless you want to modify the package. If you just want to use the Stripe Python bindings, you should run:

    pip install --upgrade git+ssh://github.com/Downtownapp/aiostripe

See http://www.pip-installer.org/en/latest/index.html for instructions on installing pip. If you're not using virtualenv, you may have to prefix those commands with `sudo`. You can learn more about virtualenv at http://www.virtualenv.org/

To install from source, run:

    python setup.py install

## Documentation

Please see https://stripe.com/docs/api/python for the most up-to-date documentation.

## Testing

We commit to being compatible with Python 3.5+.  We need to test against all of these environments to ensure compatibility.  Travis CI will automatically run our tests on push.  For local testing, we use [tox](http://tox.readthedocs.org/) to handle testing across environments.

### Setting up tox

In theory, you should be able to `pip install tox` and then simply run `tox` from the project root. In reality, Tox can take a bit of finagling to get working.

You'll need an interpreter installed for each of the versions of python we test (see the envlist in tox.ini). You can find these releases [on the Python site](https://www.python.org/download/releases). If you're using OS X, it may be easier to get Python 3.5 from Homebrew with `brew install python`.

You may choose not to go through the hassle of installing interpreters for every Python version we support (because we only support Python 3.5). It's useful to test at least one Python 3.x but can generally rely on Travis to find edge cases with other interpreters. You can test a specific interpreter such as Python 3.5 with `tox -e py35`.

The system Python on OS X has been known to cause issues. You'll probably want to `brew install python` or equivalent.  If tox complains about `pkg_resources.DistributionNotFound` then some of your Python libraries are probably still linked to the system python installation.  To fix this for virtualenv you should `sudo pip uninstall virtualenv; pip install virtualenv`.

### Running specific tests

You can specify a module, TestCase or single test to run by passing it as an argument to tox.  For example, to run only the `test_save` test of the `UpdateableAPIResourceTests` case from the `test_resources` module on Python 3.5:

    tox -e py35 -- --test-suite stripe.test.test_resources.UpdateableAPIResourceTests.test_save


CHANGELOG
=== 2.0.0 2015-12-01

* Rewritten to make use of aiohttp and asyncio

=== 1.27.1 2015-10-02

* Official Python 3.4 & 3.5 compatibility
* Add configurable HTTP client
* Add ability to delete attributes

=== 1.27.0 2015-09-14

* Products, SKUs, Orders resources

=== 1.26.0 2015-09-11

* Add support for new 429 rate limit response

=== 1.25.0 2015-08-17

* Added refund listing, creation and retrieval

=== 1.24.1 2015-08-05

* Fix error handling for Python 2.6

=== 1.24.0 2015-08-03

* Managed accounts can now be deleted
* Added dispute listing and retrieval

=== 1.23.0 2015-07-06

* Include response headers in exceptions

=== 1.22.3 2015-06-04

* Fix saving `additional_owners` on managed accounts

=== 1.22.2 2015-04-08

* Fix saving manage accounts

=== 1.22.1 2015-03-30

* Pass `stripe_account` to Balance.retrieve

=== 1.22.0 2015-03-22

* Added methods for updating and saving arrays of objects

=== 1.21.0 2015-02-19

* Added Bitcoin Receiver update and delete methods

=== 1.20.2 2015-01-21

* Remove support for top-level bitcoin transactions

=== 1.20.1 2015-01-07

* Adding bitcoin receiver and transaction objects

=== 1.20.0 2014-12-23

* Adding support for file uploads resource

=== 1.19.1 2014-10-23

* Remove redundant manual SSL blacklist preflight check

=== 1.19.0 2014-07-26

* Application Fee refunds now a list instead of array

=== 1.18.0 2014-06-17

* Add metadata for disputes and refunds

=== 1.17.0 2014-06-10

* Remove official support for Python 2.5

=== 1.16.0 2014-05-28

* Support for canceling transfers

=== 1.15.1 2014-05-21

* Support cards for recipients.

=== 1.14.1 2014-05-19

* Disable loading the ssl module on the Google App Engine dev server.

=== 1.14.0 2014-04-09

* Use DER encoded certificate for checksumming
* Don't rely on SNI support in integration tests

=== 1.13.0 2014-04-09

* Update bundled ca-certificates
* Add certificate blacklist for CVE-2014-0160 mitigation

=== 1.12.2 2014-03-13

* Fix syntax errors in setup.py metadata

=== 1.12.1 2014-03-13

* Added license and other metadata in setup.py
* Fix `__repr__` in Python 3
* Support pickling of responses

=== 1.12.0 2014-01-29

* Added support for multiple subscriptions per customer

=== 1.11.0 2013-12-05

* Added extensive unit tests
* Extended functional test coverage
* Refactored code into modules and out of stripe/__init__.py
* Abstracted http library selection and use from the `APIRequestor` into `stripe.http_client`
* Refactored `StripeObject` to inherit from `dict` and avoid direct access of `__dict__`.
* PEP8ified the codebase and enforced with a test.
* Proper encoding of timezone aware datetimes

Backwards Incompatible Changes

* The `to_dict` and `values` methods on resources no longer recursively convert objects to plain `dict`s.  All resources now inherit from `dict` but are functionally different in that you cannot set a value to an empty string and cannot delete items.
* The `previous_metadata` attribute on resources is now protected.
* Timezone aware `datetime` objects passed to the API will now be encoded in a way that does not depend on the local system time.  If you are passing timezone-aware datetimes to our API and your server time is not already in UTC, this will change the value passed to our API.

=== 1.10.8 2013-12-02

* Add stripe.ApplicationFee resource

=== 1.9.8 2013-10-17

* Removed incorrect test.

=== 1.9.7 2013-10-10

* Add support for metadata.

=== 1.9.6 2013-10-08

* Fix issue with support for closing disputes.

=== 1.9.5 2013-09-18

* Add support for closing disputes.

=== 1.9.4 2013-08-13

* Add stripe.Balance and stripe.BalanceTransaction resources

=== 1.9.3 2013-08-12

* Add support for unsetting attributes by setting to None.
  Setting properties to a blank string is now an error.

=== 1.9.2 2013-07-12

* Add support for multiple cards API

=== 1.9.1 2013-05-03

* Remove 'id' from the list of permanent attributes

=== 1.9.0 2013-04-25

* Support for Python 3 (github issue #32)

=== 1.8.0 2013-04-11

* Allow transfers to be creatable
* Add new stripe.Recipient resource

=== 1.7.10 2013-02-21

* Add 'id' to the list of permanent attributes

=== 1.7.9 2013-02-01

* Add support for passing options when retrieving Stripe objects
  e.g., stripe.Charge.retrieve("foo", params={"expand":["customer"]})

=== 1.7.8 2013-01-15

* Add support for setting a Stripe API version override.

=== 1.7.7 2012-12-18

* Update requests version check to work with requests 1.x.x (github
  issue #24)

=== 1.7.6 2012-11-08

* Add support for updating charge disputes

=== 1.7.5 2012-10-30

* Add support for creating invoices
* Add support for new invoice lines return format
* Add support for new List objects

=== 1.7.4 2012-08-31

* Add update and pay methods for Invoice resource

=== 1.7.3 2012-08-15

* Add new stripe.Account resource

* Remove uncaptured_charge tests (this has been deprecated from the API).

=== 1.7.2 2012-05-31

* Fix a bug that would cause nested objects to be mis-rendered in
  __str__ and __repr__ methods (github issues #17, #18)

=== 1.7.1 2012-05-21

* Prefer App Engine's urlfetch over requests, as that's the only thing
  that will work in App Engine's environment. Previously, if requests
  was available in the App Engine environment, we would attempt to use
  it.

=== 1.7.0 2012-05-17

* Add new delete_discount method to stripe.Customer

* Add new stripe.Transfer resource

* Switch from using HTTP Basic auth to Bearer auth. (Note: Stripe will
  support Basic auth for the indefinite future, but recommends Bearer
  auth when possible going forward)

* Numerous test suite improvements

=== 1.6.1 2011-09-14

* Parameters with value None are no longer included in API requests


coroutils/funcs.py
import collections
import operator
import sys
from functools import update_wrapper
from inspect import iscoroutinefunction

from coroutils.generator import async_generator, aiter, anext

_missing = object()


def wrapsync(func):
    """Turns synchronous function into asynchronous."""

    if iscoroutinefunction(func):
        return func

    async def inner(*args, **kwargs):
        return func(*args, **kwargs)

    return update_wrapper(inner, func)


async def aitersync(iterable):
    results = []

    async for x in aiter(iterable):
        results.append(x)

    return iter(results)


async def aall(iterable):
    async for element in aiter(iterable):
        if not element:
            return False
    return True


async def aany(iterable):
    async for element in aiter(iterable):
        if element:
            return True
    return False


@async_generator
async def aenumerate(sequence, start=0):
    n = start
    async for elem in aiter(sequence):
        await async_yield(n, elem)
        n += 1


@async_generator
async def afilter(function, iterable):
    if function is None:
        function = lambda x: x
    function = wrapsync(function)

    async for element in aiter(iterable):
        if await function(element):
            await async_yield(element)


@async_generator
async def afilterfalse(function, iterable):
    if function is None:
        function = lambda x: x
    function = wrapsync(function)

    async for element in aiter(iterable):
        if not await function(element):
            await async_yield(element)


@async_generator
async def amap(function, iterable, *args):
    function = wrapsync(function)

    iterables = (iterable, *args)
    async for args in azip(*iterables):
        await async_yield(await function(*args))


async def amax(iterable, key=None, default=_missing):
    if key is None:
        key = lambda x: x
    key = wrapsync(key)

    value = _missing
    kvalue = _missing

    async for x in aiter(iterable):
        kx = await key(x)

        if value is _missing:
            value = x
            kvalue = kx
        else:
            if kx > kvalue:
                value = x
                kvalue = kx

    if value is _missing:
        if default is not _missing:
            return default

        raise ValueError('amax() arg is an empty sequence')

    return value


async def amin(iterable, key=None, default=_missing):
    if key is None:
        key = lambda x: x
    key = wrapsync(key)

    value = _missing
    kvalue = _missing

    async for x in aiter(iterable):
        kx = await key(x)

        if value is _missing:
            value = x
            kvalue = kx
        else:
            if kx < kvalue:
                value = x
                kvalue = kx

    if value is _missing:
        if default is not _missing:
            return default

        raise ValueError('amin() arg is an empty sequence')

    return value


@async_generator
async def arange(*opts):
    return await async_yield_from(range(*opts))


async def areversed(seq):
    return aiter(reversed(await aitersync(seq)))


async def asorted(iterable, *args, **kwargs):
    return aiter(sorted(await aitersync(iterable), *args, **kwargs))


async def asum(iterable, start=0):
    async for x in aiter(iterable):
        start += x

    return start


@async_generator
async def azip(*iterables):
    # zip('ABCD', 'xy') --> Ax By
    sentinel = object()
    iterators = []
    for it in iterables:
        iterators.append(aiter(it))

    while iterators:
        result = []
        async for it in iterators:
            elem = await anext(it, sentinel)
            if elem is sentinel:
                return
            result.append(elem)

        await async_yield(tuple(result))


@async_generator
async def aaccumulate(iterable, func=None):
    'Return running totals'

    if func is None:
        func = operator.add
    func = wrapsync(func)

    # accumulate([1,2,3,4,5]) --> 1 3 6 10 15
    # accumulate([1,2,3,4,5], operator.mul) --> 1 2 6 24 120
    it = aiter(iterable)
    try:
        total = await anext(it)
    except StopAsyncIteration:
        return
    await async_yield(total)
    async for element in it:
        total = await func(total, element)
        await async_yield(total)


@async_generator
async def achain(*iterables):
    # chain('ABC', 'DEF') --> A B C D E F
    for it in iterables:
        async for element in aiter(it):
            await async_yield(element)


@async_generator
async def afrom_iterable(iterables):
    # chain.from_iterable(['ABC', 'DEF']) --> A B C D E F
    async for it in aiter(iterables):
        async for element in aiter(it):
            await async_yield(element)


@async_generator
async def acompress(data, selectors):
    # compress('ABCDEF', [1,0,1,0,1,1]) --> A C E F
    async for d, s in azip(data, selectors):
        if s:
            await async_yield(d)


@async_generator
async def acount(start=0, step=1):
    # count(10) --> 10 11 12 13 14 ...
    # count(2.5, 0.5) -> 2.5 3.0 3.5 ...
    n = start
    while True:
        await async_yield(n)
        n += step


@async_generator
async def acycle(iterable):
    # cycle('ABCD') --> A B C D A B C D A B C D ...
    saved = []
    async for element in aiter(iterable):
        await async_yield(element)
        saved.append(element)
    while saved:
        for element in saved:
            await async_yield(element)


@async_generator
async def adropwhile(predicate, iterable):
    # dropwhile(lambda x: x<5, [1,4,6,4,1]) --> 6 4 1
    predicate = wrapsync(predicate)
    iterable = aiter(iterable)
    async for x in iterable:
        if not await predicate(x):
            await async_yield(x)
            break
    async for x in iterable:
        await async_yield(x)


@async_generator
async def aislice(iterable, *args):
    # islice('ABCDEFG', 2) --> A B
    # islice('ABCDEFG', 2, 4) --> C D
    # islice('ABCDEFG', 2, None) --> C D E F G
    # islice('ABCDEFG', 0, None, 2) --> A C E G
    s = slice(*args)
    it = iter(range(s.start or 0, s.stop or sys.maxsize, s.step or 1))
    try:
        nexti = next(it)
    except StopIteration:
        return
    async for i, element in aenumerate(iterable):
        if i == nexti:
            await async_yield(element)
            nexti = next(it)


@async_generator
async def arepeat(value, times=None):
    # repeat(10, 3) --> 10 10 10
    if times is None:
        while True:
            await async_yield(value)
    else:
        for i in range(times):
            await async_yield(value)


@async_generator
async def astarmap(function, iterable):
    # starmap(pow, [(2,5), (3,2), (10,3)]) --> 32 9 1000
    function = wrapsync(function)
    async for args in aiter(iterable):
        await async_yield(await function(*args))


@async_generator
async def atakewhile(predicate, iterable):
    predicate = wrapsync(predicate)
    # takewhile(lambda x: x<5, [1,4,6,4,1]) --> 1 4
    async for x in aiter(iterable):
        if await predicate(x):
            await async_yield(x)
        else:
            break


@async_generator
async def atee(iterable, n=2):
    it = aiter(iterable)
    deques = [collections.deque() for i in range(n)]

    async def gen(mydeque):
        while True:
            if not mydeque:  # when the local deque is empty
                try:
                    newval = await anext(it)  # fetch a new value and
                except StopAsyncIteration:
                    return
                for d in deques:  # load it to all the deques
                    d.append(newval)
            await async_yield(mydeque.popleft())

    res = []
    for d in deques:
        res.append(await gen(d))
    return tuple(res)


class AsyncZipExhausted(Exception):
    pass


@async_generator
async def azip_longest(*args, **kwds):
    # zip_longest('ABCD', 'xy', fillvalue='-') --> Ax By C- D-
    fillvalue = kwds.get('fillvalue')
    counter = len(args) - 1

    @async_generator
    async def sentinel():
        nonlocal counter
        if not counter:
            raise AsyncZipExhausted
        counter -= 1
        await async_yield(fillvalue)

    fillers = await arepeat(fillvalue)
    iterators = []
    for it in args:
        iterators.append(await achain(it, sentinel(), fillers))
    try:
        while iterators:
            await async_yield(tuple(await aitersync(amap(anext, iterators))))
    except AsyncZipExhausted:
        pass

__all__ = ['wrapsync', 'aitersync', 'aall', 'aany', 'aenumerate', 'afilter', 'afilterfalse', 'amap', 'amax', 'amin',
           'arange', 'areversed', 'asorted', 'asum', 'azip', 'aaccumulate', 'achain', 'afrom_iterable', 'acompress',
           'acount', 'acycle', 'adropwhile', 'aislice', 'arepeat', 'astarmap', 'atakewhile', 'atee', 'azip_longest']
